### ISSUE TITLE
Qt5 QOpenGLWidget: Failed to create context

### ISSUE THREAD

**Issue by josephthlee** [rockchip-linux/libmali#72]
I am trying to use QOpenGLWidget on RK3399 device with:

Ubuntu kernel 4.4.194 #11 SMP aarch64 GNU/Linux
Using Qt version 5.9.5 in /usr/lib/aarch64-linux-gnu

I can run glmark2-es2 successfully below,

![12961632791102_ pic](https://user-images.githubusercontent.com/30864375/135005612-caea09ea-0772-4776-ba2c-bea362b32b6d.jpg)

However, when I try to use QOpenGLWidget and run my own compiled Qt5 with c++ program it shows:

```
QXcbIntegration: Cannot create platform OpenGL context, neither GLX nor EGL are enabled
QXcbIntegration: Cannot create platform OpenGL context, neither GLX nor EGL are enabled
QOpenGLWidget: Failed to create context
QXcbIntegration: Cannot create platform OpenGL context, neither GLX nor EGL are enabled
QOpenGLWidget: Failed to create context
```
I tried to export QT_QPA_PLATFORM to eglfs and linuxfb but all in vein.

Now RK3399 fully supports QGL? Or how can I use Qt5::QOpenGLXXX components successfully on it?

Appreciate your help.

**Comment by JeffyCN** [comment]
1, linuxfb doesn't support opengl[es]
2, if you are using official ubuntu/debian, the official prebuilt qt packages are not support egl for arm64(configuration not enabled)

so maybe you can try a armhf rootfs.

**Comment by JeffyCN** [comment]
BTW, using opengl[es] with x11 requires xserver to provide DRI2(check the /var/log/Xorg.0.log), which needs a custom version of xserver(if using libmali).

**Comment by josephthlee** [comment]
Thanks for the clues.

My Ubuntu is not official but downloaded rootfs.
I try to export QT_DEBUG_PLUGINS=1 and I found that an error "libEGL.so.1 not found"
And then I link libMali.so to libEGL.so.1 and it works. More specifically, no "Failed to create context" error anymore.

Your Xorg-xserver document is quite helpful. One more question, how can I watch the usage of GPU on it (like top)?

**Comment by JeffyCN** [comment]
> My Ubuntu is not official but downloaded rootfs. I try to export QT_DEBUG_PLUGINS=1 and I found that an error "libEGL.so.1 not found" at runtime. And then I link libMali.so to libEGL.so.1 and it works. More specifically, no "Failed to create context" error anymore.

if you are using debian/ubuntu, we would prefer to build libmali into .deb packages. prebuilt ones:
https://github.com/JeffyCN/rockchip_mirrors/tree/debian/packages/arm64/libmali

it would create links in /usr/lib/aarch64-linux-gnu/mali and have a ld conf for it:
cat /etc/ld.so.conf.d/00-arm64-mali.conf
/usr/lib/aarch64-linux-gnu/mali/

find /usr/lib/aarch64-linux-gnu/mali/|grep EGL
/usr/lib/aarch64-linux-gnu/mali/libEGL.so.1
/usr/lib/aarch64-linux-gnu/mali/libEGL.so


> Your Xorg-xserver document is quite helpful. Thanks for this. One more question, how can I watch the usage of GPU on it (like top)?
root@linaro-alip:~# grep "" /sys/devices/platform/*gpu/uti*
/sys/devices/platform/ff9a0000.gpu/utilisation:0 <--- usage in percent
/sys/devices/platform/ff9a0000.gpu/utilisation_period:100

**Comment by josephthlee** [comment]
Thanks for those useful information and now I can watch the GPU utilization now.

Now I think my current Mali driver is libmali-midgard-t86x-r14p0.so.

![12981632917940_ pic](https://user-images.githubusercontent.com/30864375/135267148-6e8d31e5-2540-44a1-b6d3-3520d73015f1.jpg)

I have checked the driver link you gave above but I am not familiar with the name convention on the list. I cannot find a document, so I have the following questions below:

1. I think that I can download libmali-midguard-t86x-???, what is the different between r14p0 and r18p0? Can I use r180p or it is just a different hardware model name? ( Currently mine is r14p0)
2. What is the difference between libmali-midgard-t86x-r14p0-x11-dbgsym_1.9-1_arm64.deb and libmali-midgard-t86x-r18p0-x11_1.9-1_arm64.deb? I guessed the first one is with debug symbol and the later one is stripped?
3. My Qt5 application is bring up by xinit, so the best candidate should be  libmali-midgard-t86x-r14p0-x11_1.9-1_arm64.deb right?
